File: frontend/display_app/lib/models/presentation_state.dart
import 'ws_event.dart';

/// The high-level phase the display is in. Drives the display app's screen FSM.
enum DisplayPhase {
  splash,
  advertisement,
  waiting, // showing QR, awaiting pairing
  connecting,
  loading,
  welcome,
  catalogue, // the collection / curated recommendations grid
  cart, // the saved-outfits page mirrored from the controller
  checkout, // the order review mirrored from the controller
  presenting, // a product is on screen (Presentation mode)
  thankYou,
  disconnected,
}

/// Which view of the presented product is active.
enum PresentationView { hero, gallery, video, colorways, specifications }

/// The synchronized state of the product currently on the display.
///
/// This is the single object the display renders and the mobile mirrors. Every
/// live-sync [WsEvent] produces a new immutable copy via [applyEvent].
class ProductPresentation {
  const ProductPresentation({
    required this.productId,
    this.variantId,
    this.size,
    this.view = PresentationView.hero,
    this.imageIndex = 0,
    this.zoom = 1.0,
    this.panX = 0.0,
    this.panY = 0.0,
    this.showAIHighlights = false,
    this.relatedMediaId,
    this.videoPlaying = false,
    this.videoPositionMs = 0,
    this.videoMuted = false,
    this.detailsExpanded = false,
    this.scrollFraction = 0,
  });

  final String productId;
  final String? variantId;

  /// The size the salesperson has selected (mirrored onto the display).
  final String? size;
  final PresentationView view;
  final int imageIndex;
  final double zoom;
  final double panX;
  final double panY;
  final bool showAIHighlights;
  final String? relatedMediaId;
  final bool videoPlaying;
  final int videoPositionMs;
  final bool videoMuted;

  /// Whether the full product-details panel is expanded (synced from the sheet).
  final bool detailsExpanded;

  /// Normalised scroll position (0..1) of the associate's product-detail sheet,
  /// mirrored so the display's info panel scrolls in step.
  final double scrollFraction;

  ProductPresentation copyWith({
    String? productId,
    String? variantId,
    String? size,
    PresentationView? view,
    int? imageIndex,
    double? zoom,
    double? panX,
    double? panY,
    bool? showAIHighlights,
    String? relatedMediaId,
    bool? videoPlaying,
    int? videoPositionMs,
    bool? videoMuted,
    bool? detailsExpanded,
    double? scrollFraction,
  }) => ProductPresentation(
    productId: productId ?? this.productId,
    variantId: variantId ?? this.variantId,
    size: size ?? this.size,
    view: view ?? this.view,
    imageIndex: imageIndex ?? this.imageIndex,
    zoom: zoom ?? this.zoom,
    panX: panX ?? this.panX,
    panY: panY ?? this.panY,
    showAIHighlights: showAIHighlights ?? this.showAIHighlights,
    relatedMediaId: relatedMediaId ?? this.relatedMediaId,
    videoPlaying: videoPlaying ?? this.videoPlaying,
    videoPositionMs: videoPositionMs ?? this.videoPositionMs,
    videoMuted: videoMuted ?? this.videoMuted,
    detailsExpanded: detailsExpanded ?? this.detailsExpanded,
    scrollFraction: scrollFraction ?? this.scrollFraction,
  );

  /// Reduce a realtime [event] into a new presentation state. Pure function —
  /// the same reducer runs on both apps so they stay perfectly in sync.
  ProductPresentation applyEvent(WsEvent event) {
    switch (event.type) {
      case WsEventType.showProduct:
        return ProductPresentation(
          productId: event.productId ?? productId,
          variantId: event.variantId ?? variantId,
          size: event.size ?? size,
        );
      case WsEventType.changeColor:
        return copyWith(
          variantId: event.variantId,
          size: event.size ?? size,
          imageIndex: 0,
          zoom: 1,
          panX: 0,
          panY: 0,
          view: PresentationView.hero,
        );
      case WsEventType.changeSize:
        return copyWith(size: event.size ?? size);
      case WsEventType.scrollSync:
        return copyWith(scrollFraction: event.fraction ?? scrollFraction);
      case WsEventType.changeImage:
        return copyWith(
          imageIndex: event.imageIndex ?? imageIndex,
          view: PresentationView.hero,
          zoom: 1,
          panX: 0,
          panY: 0,
        );
      case WsEventType.zoomImage:
        return copyWith(
          zoom: event.scale ?? zoom,
          panX: event.focalX ?? panX,
          panY: event.focalY ?? panY,
        );
      case WsEventType.panImage:
        return copyWith(
          panX: event.offsetX ?? panX,
          panY: event.offsetY ?? panY,
        );
      case WsEventType.resetZoom:
        return copyWith(zoom: 1, panX: 0, panY: 0);
      case WsEventType.showAIHighlights:
        return copyWith(
          showAIHighlights: (event.payload['visible'] as bool?) ?? true,
        );
      case WsEventType.showDetails:
        return copyWith(
          detailsExpanded:
              (event.payload['expanded'] as bool?) ?? !detailsExpanded,
        );
      case WsEventType.showRelatedMedia:
        return copyWith(
          relatedMediaId: event.mediaId,
          view: PresentationView.gallery,
        );
      case WsEventType.playVideo:
        return copyWith(view: PresentationView.video, videoPlaying: true);
      case WsEventType.pauseVideo:
        return copyWith(videoPlaying: false);
      case WsEventType.seekVideo:
        return copyWith(videoPositionMs: event.positionMs ?? videoPositionMs);
      case WsEventType.muteVideo:
        return copyWith(videoMuted: (event.payload['muted'] as bool?) ?? true);
      default:
        return this;
    }
  }
}

File: frontend-rn/display_app/src/models/presentationState.ts
import { WsEvent, WsEventType } from './wsEvent';

/// The high-level phase the display is in. Drives the display app's screen FSM.
export enum DisplayPhase {
  splash = 'splash',
  advertisement = 'advertisement',
  waiting = 'waiting',
  connecting = 'connecting',
  loading = 'loading',
  welcome = 'welcome',
  catalogue = 'catalogue',
  cart = 'cart',
  checkout = 'checkout',
  presenting = 'presenting',
  thankYou = 'thankYou',
  disconnected = 'disconnected',
}

/// Which view of the presented product is active.
export enum PresentationView {
  hero = 'hero',
  gallery = 'gallery',
  video = 'video',
  colorways = 'colorways',
  specifications = 'specifications',
}

/// The synchronized state of the product currently on the display. Immutable;
/// every live-sync WsEvent produces a new copy via applyEvent. Ported 1:1.
export class ProductPresentation {
  readonly productId: string;
  readonly variantId?: string;
  readonly size?: string;
  readonly view: PresentationView;
  readonly imageIndex: number;
  readonly zoom: number;
  readonly panX: number;
  readonly panY: number;
  readonly showAIHighlights: boolean;
  readonly relatedMediaId?: string;
  readonly videoPlaying: boolean;
  readonly videoPositionMs: number;
  readonly videoMuted: boolean;
  readonly detailsExpanded: boolean;
  readonly fullscreen: boolean;
  readonly scrollFraction: number;

  constructor(init: {
    productId: string;
    variantId?: string;
    size?: string;
    view?: PresentationView;
    imageIndex?: number;
    zoom?: number;
    panX?: number;
    panY?: number;
    showAIHighlights?: boolean;
    relatedMediaId?: string;
    videoPlaying?: boolean;
    videoPositionMs?: number;
    videoMuted?: boolean;
    detailsExpanded?: boolean;
    fullscreen?: boolean;
    scrollFraction?: number;
  }) {
    this.productId = init.productId;
    this.variantId = init.variantId;
    this.size = init.size;
    this.view = init.view ?? PresentationView.hero;
    this.imageIndex = init.imageIndex ?? 0;
    this.zoom = init.zoom ?? 1.0;
    this.panX = init.panX ?? 0.0;
    this.panY = init.panY ?? 0.0;
    this.showAIHighlights = init.showAIHighlights ?? false;
    this.relatedMediaId = init.relatedMediaId;
    this.videoPlaying = init.videoPlaying ?? false;
    this.videoPositionMs = init.videoPositionMs ?? 0;
    this.videoMuted = init.videoMuted ?? false;
    this.detailsExpanded = init.detailsExpanded ?? false;
    this.fullscreen = init.fullscreen ?? false;
    this.scrollFraction = init.scrollFraction ?? 0;
  }

  copyWith(patch: Partial<{
    productId: string;
    variantId: string;
    size: string;
    view: PresentationView;
    imageIndex: number;
    zoom: number;
    panX: number;
    panY: number;
    showAIHighlights: boolean;
    relatedMediaId: string;
    videoPlaying: boolean;
    videoPositionMs: number;
    videoMuted: boolean;
    detailsExpanded: boolean;
    fullscreen: boolean;
    scrollFraction: number;
  }>): ProductPresentation {
    return new ProductPresentation({
      productId: patch.productId ?? this.productId,
      variantId: patch.variantId ?? this.variantId,
      size: patch.size ?? this.size,
      view: patch.view ?? this.view,
      imageIndex: patch.imageIndex ?? this.imageIndex,
      zoom: patch.zoom ?? this.zoom,
      panX: patch.panX ?? this.panX,
      panY: patch.panY ?? this.panY,
      showAIHighlights: patch.showAIHighlights ?? this.showAIHighlights,
      relatedMediaId: patch.relatedMediaId ?? this.relatedMediaId,
      videoPlaying: patch.videoPlaying ?? this.videoPlaying,
      videoPositionMs: patch.videoPositionMs ?? this.videoPositionMs,
      videoMuted: patch.videoMuted ?? this.videoMuted,
      detailsExpanded: patch.detailsExpanded ?? this.detailsExpanded,
      fullscreen: patch.fullscreen ?? this.fullscreen,
      scrollFraction: patch.scrollFraction ?? this.scrollFraction,
    });
  }

  /// Reduce a realtime event into a new presentation state. Pure function —
  /// the same reducer runs on both apps so they stay perfectly in sync.
  applyEvent(event: WsEvent): ProductPresentation {
    switch (event.type) {
      case WsEventType.showProduct:
        return new ProductPresentation({
          productId: event.productId ?? this.productId,
          variantId: event.variantId ?? this.variantId,
          size: event.size ?? this.size,
        });
      case WsEventType.changeColor:
        return this.copyWith({
          variantId: event.variantId,
          size: event.size ?? this.size,
          imageIndex: 0,
          zoom: 1,
          panX: 0,
          panY: 0,
          view: PresentationView.hero,
        });
      case WsEventType.changeSize:
        return this.copyWith({ size: event.size ?? this.size });
      case WsEventType.scrollSync:
        return this.copyWith({ scrollFraction: event.fraction ?? this.scrollFraction });
      case WsEventType.changeImage:
        return this.copyWith({
          imageIndex: event.imageIndex ?? this.imageIndex,
          view: PresentationView.hero,
          zoom: 1,
          panX: 0,
          panY: 0,
        });
      case WsEventType.zoomImage:
        return this.copyWith({
          zoom: event.scale ?? this.zoom,
          panX: event.focalX ?? this.panX,
          panY: event.focalY ?? this.panY,
        });
      case WsEventType.panImage:
        return this.copyWith({
          panX: event.offsetX ?? this.panX,
          panY: event.offsetY ?? this.panY,
        });
      case WsEventType.resetZoom:
        return this.copyWith({ zoom: 1, panX: 0, panY: 0 });
      case WsEventType.showAIHighlights:
        return this.copyWith({
          showAIHighlights: (event.payload.visible as boolean | undefined) ?? true,
        });
      case WsEventType.fullscreen:
        return this.copyWith({ fullscreen: event.isFullscreen ?? !this.fullscreen });
      case WsEventType.showDetails:
        return this.copyWith({
          detailsExpanded: (event.payload.expanded as boolean | undefined) ?? !this.detailsExpanded,
        });
      case WsEventType.showRelatedMedia:
        return this.copyWith({
          relatedMediaId: event.mediaId,
          view: PresentationView.gallery,
        });
      case WsEventType.playVideo:
        return this.copyWith({ view: PresentationView.video, videoPlaying: true });
      case WsEventType.pauseVideo:
        return this.copyWith({ videoPlaying: false });
      case WsEventType.seekVideo:
        return this.copyWith({ videoPositionMs: event.positionMs ?? this.videoPositionMs });
      case WsEventType.muteVideo:
        return this.copyWith({ videoMuted: (event.payload.muted as boolean | undefined) ?? true });
      default:
        return this;
    }
  }
}
